web-crawler/src/main/java/com/google/sps/webcrawler/WebCrawler.java

```diff
@@ -0,0 +1,231 @@
+  //   } catch (InterruptedException e) {}
+  // }
+
+  // @TODO [Test with Google Custom Search. Extract other metadata.]
+  /** 
+   * Searches for {@code candidateName} in the Google Custom Search engine and finds URLs of news
+   * articles.
+   *
+   * @see <a href=
+   *    "https://hc.apache.org/httpcomponents-client-ga/httpclient/examples/org/apache/" +
+   *    "http/examples/client/ClientWithResponseHandler.java">Code reference</a>
+   */
+  public List<String> getUrlsFromCustomSearch(String candidateName) {
+    List<String> urls;
+    String request =
+        String.format("https://www.googleapis.com/customsearch/v1?key=%s&cx=%s&q=%s",
+            CUSTOM_SEARCH_KEY, CUSTOM_SEARCH_ENGINE_ID, candidateName.replace(" ", "%20"));
+    CloseableHttpClient httpclient = HttpClients.createDefault();
+    try {
+      HttpGet httpGet = new HttpGet(request);
+      ResponseHandler<String> responseHandler = new ResponseHandler<String>() {
+        @Override
+        public String handleResponse(final HttpResponse response) throws IOException {
+          int status = response.getStatusLine().getStatusCode();
+          if (status >= 200 && status < 300) {
+              HttpEntity entity = response.getEntity();
+              return entity != null ? EntityUtils.toString(entity) : null;
+          } else {
+              throw new ClientProtocolException("Unexpected response status: " + status);
+          }
+        }
+      };
+      String responseBody = httpclient.execute(httpGet, responseHandler);
+      httpclient.close();
+      Gson gson = new Gson();
+      Object jsonResponse = gson.fromJson(responseBody, Object.class);
+      //@TODO [Unpack {@code jsonResponse} and find URLs.]
+      return Arrays.asList(
+          "https://www.cnn.com/2020/06/23/politics/aoc-ny-primary-14th-district/index.html");
+    } catch (IOException e){
+      System.out.println("[ERROR] Error occurred with fetching URLs from Custom Search: " + e);
+      return Arrays.asList(
+          "https://www.cnn.com/2020/06/23/politics/aoc-ny-primary-14th-district/index.html");
+    }
+  }
+
+  /** 
+   * Checks robots.txt for permission to web-scrape, scrapes webpage if permitted and extracts
+   * textual content.
+   */
+  public NewsArticle scrapeAndExtractHtml(String url) {
+    String robotsUrl = url.substring(0, url.indexOf("/", url.indexOf("//") + 2) + 1)
+        + "robots.txt";
+    try {
+      InputStream robotsTxtStream = new URL(robotsUrl).openStream();
+      RobotsTxt robotsTxt = RobotsTxt.read(robotsTxtStream);
+      String webpagePath = url.substring(url.indexOf("/", url.indexOf("//") + 2));
+      Grant grant = robotsTxt.ask("*", webpagePath);
+      // Check and set crawl delay.
+      if (grant != null && grant.getCrawlDelay() != null) {
+        if (nextAccessTimes.containsKey(robotsUrl)) {
+          // Wait until the crawl delay fully passes.
+          if (System.currentTimeMillis() < nextAccessTimes.get(robotsUrl)) {
```

Comment:
I don't like these nested loops. Consider turning all of this into a helper function that does if this && that && those